### ISSUE TITLE
jaeger-collector fails to send spans to kafka

### ISSUE THREAD

**Issue by dasn-TeraBundle** [jaegertracing/jaeger#3687]
**Describe the bug**
jaeger-collector fails to send spans to Kafka when one of the leader broker of some partition is restarted

`{"level":"error","ts":1650947724.119294,"caller":"kafka/writer.go:60","msg":"kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.","stacktrace":"github.com/jaegertracing/jaeger/plugin/storage/kafka.NewSpanWriter.func2\n\t/builds/ugbu-dev/jaeger/jaeger/plugin/storage/kafka/writer.go:60"}
`

**To Reproduce**
1.  Setup a 3 broker Kafka cluster with partition count as 10 and replica as 2
2. Check the leader broker and scale down that. 
3. Wait for sometime and then again scale up the stopped broker.
4. Follow this sequence multiple times to observe the issue

**Expected behavior**
jaeger-collector sending spans with no errors.

**Screenshots**

**Version (please complete the following information):**
 - OS: [e.g. Linux]
 - Jaeger version:   - 1.21
 - Kafka version - 2.8.0
 - Deployment: - Kubernetes
 

**What troubleshooting steps did you try?**
1.  Zookeeper restarted
2. All 3 Kafka brokers restarted
It started working again after deleting the jaeger-spans topic in Kafak

**Additional context**

**Comment by juame** [comment]
Similar case with link to another issue #3578:

Two Collector (jaegertracing/jaeger-collector:1.34.1) pods (similar logs):
```
# 8 times:
{"level":"error","ts":1656905269.36191,"caller":"kafka/writer.go:61","msg":"kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.","stacktrace":"github.com/jaegertracing/jaeger/plugin/storage/kafka.NewSpanWriter.func2\n\tgithub.com/jaegertracing/jaeger/plugin/storage/kafka/writer.go:61"}
...
...
# 3 times:
{"level":"error","ts":1656905279.2345138,"caller":"kafka/writer.go:61","msg":"kafka server: In the middle of a leadership election, there is currently no leader for this partition and hence it is unavailable for writes.","stacktrace":"github.com/jaegertracing/jaeger/plugin/storage/kafka.NewSpanWriter.func2\n\tgithub.com/jaegertracing/jaeger/plugin/storage/kafka/writer.go:61"}
...
...
# 19 times:
{"level":"error","ts":1656905287.1972206,"caller":"kafka/writer.go:61","msg":"circuit breaker is open","stacktrace":"github.com/jaegertracing/jaeger/plugin/storage/kafka.NewSpanWriter.func2\n\tgithub.com/jaegertracing/jaeger/plugin/storage/kafka/writer.go:61"}
...
# 5 times:
{"level":"error","ts":1656905298.948992,"caller":"kafka/writer.go:61","msg":"kafka server: Tried to send a message to a replica that is not the leader for some partition. Your metadata is out of date.","stacktrace":"github.com/jaegertracing/jaeger/plugin/storage/kafka.NewSpanWriter.func2\n\tgithub.com/jaegertracing/jaeger/plugin/storage/kafka/writer.go:61"}
...
{"level":"info","ts":1657018788.6436694,"caller":"flags/service.go:154","msg":"Shutting down"}
{"level":"info","ts":1657018788.6437225,"caller":"healthcheck/handler.go:129","msg":"Health Check state change","status":"unavailable"}
{"level":"info","ts":1657018788.6437998,"caller":"channelz/funcs.go:340","msg":"[core][Server #1 ListenSocket #2] ListenSocket deleted","system":"grpc","grpc_log":true}
{"level":"info","ts":1657018788.6451657,"caller":"healthcheck/handler.go:129","msg":"Health Check state change","status":"unavailable"}
{"level":"info","ts":1657018788.6466353,"caller":"healthcheck/handler.go:129","msg":"Health Check state change","status":"unavailable"}
```

One Ingester (jaegertracing/jaeger-ingester:1.34.1) pod (container crashing continuously: #3578):
```
{"level":"info","ts":1656905373.8032048,"caller":"consumer/consumer.go:167","msg":"Starting error handler","partition":0}
{"level":"info","ts":1656905373.8032458,"caller":"consumer/consumer.go:110","msg":"Starting message handler","partition":0}
{"level":"info","ts":1656905373.8129237,"caller":"consumer/processor_factory.go:65","msg":"Creating new processors","partition":0}

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0xdcee46]

goroutine 1960977 [running]:
github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/dbmodel.FromDomain.convertProcess(...)
	github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/dbmodel/from_domain.go:123
github.com/jaegertracing/jaeger/plugin/storage/es/spanstore/dbmodel.FromDomain.convertSpanEmbedProcess(...
...
```

While the collector was still receiving spans (`jaeger_collector_spans_received_total` increasing, `jaeger_collector_spans_rejected_total=0` and `jaeger_collector_spans_dropped_total=0`) no spans were written into the Kafka topic anymore. Only after a restart of all collector pods (today), the messages were written into the Kafka topic again.

We rolled out a custom image of ingester to overcome the panic (based on 1.34.1):
```
{"level":"warn","ts":1656912269.8027709,"caller":"spanstore/writer.go:153","msg":"invalid span received, skipping...","error":"nil span or span.Process","span":"trace_id:<"}
{"level":"warn","ts":1656912269.8031654,"caller":"spanstore/writer.go:153","msg":"invalid span received, skipping...","error":"nil span or span.Process","span":"trace_id:<"}
```

Patch for `plugin/storage/es/spanstore/writer.go`:
```
func (s *SpanWriter) WriteSpan(_ context.Context, span *model.Span) error {
	if span == nil || span.Process == nil {
		s.logger.Warn("invalid span received, skipping...", zap.Any("error", errors.New("nil span or span.Process")), zap.Any("span", span))
		return nil
	}
	...
	...
}
```

**Comment by juan-ramirez-sp** [comment]
@juame Has a PR been drafted with your changes? I'm interested in getting this fixed for my own deployment and I'd like for your change to be reviewed. 

@yurishkuro I recall you mention that you didn't want this kinda logic on the consumer side, but can we reconsider due to how disruptive this bug is?

**Comment by jkowall** [comment]
Should be fixed in Jaeger v2 when we have moved to otel, feel free to re-open if not.
